text/0038-scrubbing-sensitive-data.md

```diff
@@ -0,0 +1,136 @@
+
+_Pros:_
+
+- Relay would not have to reverse-engineer the semantics of the information supplied by the SDK.
+
+_Cons:_
+
+- Could be complex for nested JSON objects or for "Array of objects" kind of data.
+
+### Option C): Relay identifies what kind of data is present and parses it.
+
+Relay can try for each field to "guess" what kind of data it is. Guessing can be done by looking at what field in general we are guessing (field X has always a SQL query in it), or the span.op or other fields as well as the content itself. If the content is a SQL query, JSON object, Elasticsearch Query, GraphQL Query, URL, ... When Relay is certain that it knows the content is of a specific kind, it can then run a parser on it to be able to scrub values of sensitive fields.
+
+An existing example for this is [parsing URL query parameters into a separate field](https://github.com/getsentry/relay/blob/c2e666d1728a2882b82e70fdbb02192c4cb0b50a/relay-general/src/store/normalize/request.rs#L29-L48), which Relay does when normalizing the [Request Interface](https://develop.sentry.dev/sdk/event-payloads/request/).
+
+For the performance issue detection `sentry` does something similar:
+
+- https://github.com/getsentry/sentry/blob/68e44ed3e8343a5e69d0b0a51ad65c02ae427cd0/src/sentry/spans/grouping/strategy/base.py#L186
+- https://github.com/getsentry/sentry/blob/68e44ed3e8343a5e69d0b0a51ad65c02ae427cd0/src/sentry/spans/grouping/strategy/base.py#L142-L150
+
+OpenTelementry has semantic conventions for tracing. A defined set of attributes set to the span describes the span data in more detail. We could borrow those semantic conventions and add them to `span.data` so that Relay can better parse the `span.description`: https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/
+
+_Pros:_
+
+-
+
+_Cons:_
+
+- Could be expensive to try multiple guesses before the right kind of data is identified. (Maybe its SQL? no. Maybe JSON? no. So it is a URL? yes.)
+
+### Option D): Generic tozenization in Relay.
+
+Have a generic tokenizer in Relay that can not parse full fledged SQL, but can extract key/value pairs out of almost everything. With this the values of keys with potential sensitive information can be removed.
+
+_Pros:_
+
+-
+
+_Cons:_
+
+-
+
+### NEW! Option E): Improved regexes
+
+Keep the logic on how data scrubbing is done right now the same, but improve the regexes to be more specific. Especially the "password regex" could be changed to the `auth` rule does ONLY match `auth` but NOT `author` or `authorize`.
+
+With this we could add data scrubbing back to `span.description` (and potentially other fields that ware marked with `pii=maybe` right now).
+
+_Pros:_
+
+-
+
+_Cons:_
+
+-
+
+# Drawbacks
+
+(none)
```

Comment:
We should at least mention that this could have an impact on the product.

Reply:
yes, adding/removing scrubbing on spans can have impact on performance issue grouping at the very least

Reply:
I already checked with the performance issues team and they think it should not have an impact at all: 

![Screenshot 2022-11-23 at 09 42 41](https://user-images.githubusercontent.com/202325/203503295-b1ab839b-fe34-43bc-b521-42573a9b02f6.png)

Reply:
@cleptric what impacts to you have in mind?

Reply:
This might degrade the usefulness of the product or could have unknown impact if we accidentally scrub too much.